wgpu/src/util/mod.rs

```diff
@@ -76,17 +87,46 @@ pub fn make_spirv_raw(data: &[u8]) -> Cow<'_, [u32]> {
 
     // Before checking if the data starts with the magic, check if it starts
     // with the magic in non-native endianness, own & swap the data if so.
-    if words[0] == MAGIC_NUMBER.swap_bytes() {
+    if words[0] == SPIRV_MAGIC_NUMBER.swap_bytes() {
         for word in Cow::to_mut(&mut words) {
             *word = word.swap_bytes();
         }
     }
 
-    assert_eq!(
-        words[0], MAGIC_NUMBER,
-        "wrong magic word {:x}. Make sure you are using a binary SPIRV file.",
-        words[0]
-    );
+    verify_spirv_magic(&words);
+
+    words
+}
+
+/// Version of `make_spirv_raw` used for implementing [`include_spirv!`] and [`include_spirv_raw!`] macros.
+///
+/// Not public API. Also, don't even try calling at runtime; you'll get a stack overflow.
+///
+/// [`include_spirv!`]: crate::include_spirv
+#[doc(hidden)]
+pub const fn make_spirv_const<const IN: usize, const OUT: usize>(data: [u8; IN]) -> [u32; OUT] {
+    #[repr(align(4))]
+    struct Aligned<T: ?Sized>(T);
+
+    check_spirv_len(&data);
+
+    // NOTE: to get around lack of generic const expressions
+    assert!(IN / 4 == OUT);
+
+    let aligned = Aligned(data);
+    let mut words: [u32; OUT] = unsafe { mem::transmute_copy(&aligned) };
```

Comment:
[not actionable] taking this much memory on the stack here is a bit of an annoying wrinkle that leads to the stack overflow note above for calling it at runtime :/. Don't have a better idea either unfortunately that would guarantee the alignment for sure.
Well, good the method is hidden...

Reply:
Yeah, unfortunately, const stuff like this can't really be done any other way due to how limited of a set of operations we have. I initially inlined some of this into the macros because it avoided this, but I think that having less code in the macro and more in this method is better, since it's actually formatted properly.